### PR TITLE
Refactors Provider setup to not extend on instance

### PR DESCRIPTION
**Pull request by fabiokr**
Explicitelly require inheriting from a Base encryptor to make sure it
implements the base methods instead of adding those to the instance on
runtime.

### CODE REVIEW

**Comment by jmazzi** [comment]
@fabiokr this would need to be a major release because it would be a possibly breaking change for anyone who's built a customer provider. 

Would you mind having it generate a fatal error if the encryptor class has not inherited from Base?

**Comment by fabiokr** [comment]
@jmazzi How about f3de00413d441902af89bf18ae333955f9a0fac1? Also bumped version on 7c0615e7341768912d683198840e11c0c0ab9666

File: lib/crypt_keeper/model.rb [hunk 1]
```diff
@@ -127,8 +128,10 @@ def encryptor_klass
 
       # Private: Ensure that the encryptor responds to new
       def ensure_valid_encryptor!
-        unless defined?(encryptor_klass) && encryptor_klass.respond_to?(:new)
-          raise "You must specify a valid encryptor `crypt_keeper :encryptor => :aes`"
+        unless defined?(encryptor_klass) && encryptor_klass.respond_to?(:new) &&
+          encryptor_klass.superclass == Provider::Base
```
Comment (itspriddle) [inline lib/crypt_keeper/model.rb]: Instead of checking the superclass, can we check that it responds to `dump` and `load`? The class shouldn't matter IMO.

**Comment by itspriddle** [comment]
Can you explain again the issue you hit with `extend`? Is that something _any_ Rails 5 user is going to face?

I'm wondering if we could somehow handle that conditionally when necessary to avoid anyone with custom code having issues.  Maybe something along these lines (pseudocode):

```
def encryptor_klass
  if !@encryptor_klass.respond_to?(:load) && !@encryptor_klass.respond_to?(:dump)
    @encryptor_klass.include CryptKeeper::Helper::Serializer
  end
  @encryptor_klass
end
```

If we absolutely must eliminate it, I think it should land in a 1.0.0.beta1 tag or something. It's high time we had a 1.0 release and this change definitely warrants a major bump.

**Comment by itspriddle** [comment]
Ugh sorry for closing. Blame the tiny hands

File: lib/crypt_keeper/model.rb (hunk 1 above)
Comment (fabiokr) [inline lib/crypt_keeper/model.rb]: 0e8db5bf2f0f40de5ca98bf292e8b42ddbe7d4bc

**Comment by fabiokr** [comment]
@itspriddle This is where the issue is triggered: https://github.com/pressednet/backstage/pull/2689/commits/c9d05c0c15a1f951b4636c8f0f1285332607f778

Without that fix, what happens is that the spec fails with an error like "`load` method is private" when calling `provider.load`. It happens because the instance does have the CryptKeeper implementation anymore (which is public), and it starts using a `load` method implemented by ActiveSupport, which is private. It seems that activerecord is doing something that causes the instance class to be reloaded (I think some serialization code), and then the `extend` portion is lost during this.

In any case, I'm thinking that expecting the class to implement those would be more clear then extending on the instance. I'm fine with targeting a 1.0.0 release.

**Pull request merged**